### DISCUSSION TITLE
Inlcuding Image/ImageRef's in PDF output?

### DISCUSSION THREAD

**Discussion by planted-sam** [mozman/ezdxf#951] [Q&A]
Quick Context: I'm working to make a feature that will export a construction site plan as both a DXF and a PDF, and part of my content (e.g. company logos, rasterized satellite imagery) is made up of images. 

I'm able to place the images on the paperspace layout as expected, and I see the images appear when I open the DXF files (or I do when I open the DXF with the source image in the same folder), but when I use the PyMuPDF backend to generate a PDF (using `get_pdf_bytes`), I don't see the image shown in the resulting PDF, though other entities appear as expected.

Is there a straightforward way to include raster images in the PDF generation that I'm missing? Or is this something where I'd need to implement some custom functionality in order to get raster images embedded in my PDF?

Thanks for the help!

Side Note: What's the best practice for sending a DXF with image references to someone else to ensure when they open it they see the images? Zip up the DXF and all the images together? (I assume this is different dependent on what viewer the person is using)

**Comment by mbway** [comment]
it is a limitation of the current drawing addon (that should be documented [here](https://ezdxf.readthedocs.io/en/stable/addons/drawing.html) but isn't) that IMAGEDEF/IMAGE is not supported. But the only reason for this was that no-one has required support yet. You should be seeing `skipped entity` messages from the frontend if the `ezdxf` logging level has at least the INFO level visible.

If you would like to add support you would need to
- implement the deconstruction from DXF properties into a primitive representation of an image in the [frontend](https://github.com/mozman/ezdxf/blob/master/src/ezdxf/addons/drawing/frontend.py)
    - (eg RGBA array with a Matrix44 transform in a method called `draw_image_entity`).
    - I'm not sure if an IMAGEDEF without an associated IMAGE is valid, but that might be good to support as well if CAD applications can handle it
    - there are several post-processing effects like brightness that an [IMAGE](https://ezdxf.readthedocs.io/en/stable/dxfentities/image.html) entity can perform. Those can be applied in the frontend
- Add a method for rendering transformed bitmaps (eg called `draw_bitmap`) in the [backend base class](https://github.com/mozman/ezdxf/blob/master/src/ezdxf/addons/drawing/backend.py)
- implement `draw_bitmap` it either just for the pdf backend you are interested in or all of them.
    - [show_pdf_page()](https://pymupdf.readthedocs.io/en/latest/recipes-images.html#how-to-add-images-to-a-pdf-page) looks like the best way for the pdf backend. You would need to either do basic transform handling (eg only render if the transform is translation, rotation and scale which you extract from the matrix separately, or I suppose you would need to re-sample a transformed image if you want to handle 3D or non-rigid transformation
        - I'm not sure if DXF supports arbitrary IMAGE transform, I'm going by the fact there is an `Image.transform(Matrix44)` method in ezdxf

**Reply by mbway** [reply]
I can't promise anything, but if you prefer I could have a go at implementing this

**Reply by planted-sam** [reply]
I think that would be helpful, I started taking a crack at it but I think my method is a little less full featured.

My implementation was to grab the image entity's ImageRef, then grab the filename for the image from there in the frontend `draw_image_entity`, as well as the bounding box from Image's `boundary_path_wcs`. My thought was to then use the `insert_image()`[command from pymupdf](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_image), passing through the filename and transforming the boundary_path_wcs output to pass in for the `rect` parameter. Granted, I'm still in the process of building this out so I'm not sure if it works yet haha.

But the points you mention around handling post-processing does make it seem like it would be more prudent to get the raw rgba array (even if I personally don't care about that for my implementation lol).

I'd appreciate it if you want to take a go at implementing it, I think you may be more aware of the edge cases I might ignore in just trying to get my specific case to work.

**Comment by mbway** [comment]
An update on this: I have been waiting for a colleague to put together a test file involving images (I don't personally have access to autocad). They have now created the file so I should be able to make a start in the coming weeks

**Reply by planted-sam** [reply]
Ok awesome, that's good to hear. My own work priorities have shifted away from this for now, but I think I'll be back on this as well in a week or two.

**Comment by mbway** [comment]
A pull request is now ready that implements Image support in the drawing addon. https://github.com/mozman/ezdxf/pull/999

**Comment by mbway** [comment]
and support for the mupdf backend was also added https://github.com/mozman/ezdxf/pull/1001

in addition, the `ezdxf draw ...` cli command also now supports mupdf with
```
ezdxf draw <my_dxf> -o <image_path> --backend mupdf
```
https://github.com/mozman/ezdxf/pull/1000

**Comment by planted-sam** [comment]
Thanks so much for the help on this feature @mbway I appreciate your hard work!
